github.com/cockroachdb/pebble/internal/metamorphic: TestMeta failed

github.com/cockroachdb/pebble/internal/metamorphic.TestMeta [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_Pebble_Metamorphic/9761066?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_Pebble_Metamorphic/9761066?buildTab=artifacts#meta) on refs/heads/master @ [f87a15b23a26](https://github.com/cockroachdb/pebble/commits/f87a15b23a26):


```
        batch48 = db.NewIndexedBatch() #6049
        // INFO: [JOB 191] MANIFEST deleted 000223
        batch48.SingleDelete("veljlxiw@11", false /* maybeReplaceDelete */) // <nil> #6050
        // INFO: [JOB 195] MANIFEST created 000243
        // INFO: [JOB 198] ingesting: sstable created 000244
        // INFO: [JOB 195] flushed 1 memtable to L0 [000235 000237 000238 000239 000240 000241] (6.1 K), in 0.0s (0.0s total), output rate 1.3 M/s
        // INFO: [JOB 195] MANIFEST deleted 000231
        // INFO: [JOB 198] MANIFEST created 000245
        // INFO: [JOB 200] flushing 1 ingested table
        // INFO: [JOB 198] ingested L0:000244 (1006 B)
        db.Ingest(batch48) // <nil> #6051
        // INFO: [JOB 198] MANIFEST deleted 000236
        // INFO: [JOB 200] flushed 1 ingested flushable L0:000233 (1.1 K) in 0.0s (0.0s total), output rate 923 K/s
        db.Delete("uqjtiweuv@8") // <nil> #6052
        db.SingleDelete("ptaisxu@8", false /* maybeReplaceDelete */) // <nil> #6053
        iter62 = db.NewIter("", "", 0 /* key types */, 8, 9, true /* use L6 filters */, "" /* masking suffix */) // <nil> #6054
        // INFO: [JOB 202] compacting(default) L0 [000229 000224 000239 000240 000235 000238 000237 000233 000244] (9.6 K) + L0 [] (0 B)
        iter62.Prev("") // [true,"zswtr@8","fhmpefu",<no range>] <nil> #6055
        iter62.Next("") // [false] <nil> #6056
        iter62.SetBounds("bhcf@8", "gfatiuhwvrq@7") // <nil> #6057
        // INFO: [JOB 204] flushing 1 memtable to L0
        iter62.SeekLT("gfatiuhwvrq@7", "") // [false] <nil> #6058
        iter62.Next("") // [false] <nil> #6059
        iter62.Next("") // [false] <nil> #6060
        iter62.Prev("") // [false] <nil> #6061
        iter62.SetBounds("lwepjosexs@12", "wmpwcfu@10") // <nil> #6062
        iter62.SeekLT("wmpwcfu@10", "") // [true,"wcibmvyuz@8","b",<no range>] <nil> #6063
        iter62.Next("") // [false] <nil> #6064
        iter62.Next("") // [false] <nil> #6065
        iter62.Next("agdbzc@9") // [invalid] <nil> #6066
        // INFO: [JOB 204] flush error: pebble: empty table
        db.RangeKeyUnset("cubsku", "otlffwzmf", "@17") // <nil> #6067
        iter62.Last() // [true,"wcibmvyuz@8","b",<no range>] <nil> #6068
        // INFO: [JOB 202] compacting: sstable created 000246
        iter62.SeekLT("evoanburrut@12", "") // [false] <nil> #6069
        // INFO: [JOB 197] compacted(default) L5 [000232] (3.1 K) + L6 [000212] (4.5 K) -> L6 [000242] (4.5 K), in 0.0s (0.0s total), output rate 8.2 M/s
        iter62.SeekLT("xqkebapq@4", "") // [true,"wcibmvyuz@8","b",<no range>] <nil> #6070
        db.Get("fcoln@5") // [""] pebble: not found #6071
        // INFO: [JOB 205] flushing 1 ingested table
        // INFO: [JOB 205] flushed 1 ingested flushable L0:000234 (1.0 K) in 0.0s (0.0s total), output rate 58 M/s
        // INFO: [JOB 202] MANIFEST created 000247
        // INFO: [JOB 202] compacted(default) L0 [000229 000224 000239 000240 000235 000238 000237 000233 000244] (9.6 K) + L0 [] (0 B) -> L0 [000246] (2.0 K), in 0.0s (0.0s total), output rate 610 K/s
        db.Set("oidzspb@5", "bqitahuguf") // <nil> #6072
        iter62.SeekPrefixGE("wjqbc@3") // [false] <nil> #6073
        // FATAL: checker failed with error: found InternalKey eacqieb@2#1840,DEL in L0.1: fileNum=000246 and InternalKey eacqieb@2#1856,DEL in L0.0: fileNum=000234
        iter62.SeekGE("fcoln@11", "vnaqy@10") // [invalid] <nil> #6074
        // INFO: [JOB 202] MANIFEST deleted 000243
        --- FAIL: TestMeta/execution/random-025 (0.44s)
=== RUN   TestMeta/execution
    --- FAIL: TestMeta/execution (0.00s)
```
<details><summary>Help</summary>
<p>
To reproduce, try:

```bash
go test -tags 'invariants' -exec 'stress -p 1' -timeout 0 -test.v -run TestMeta$ ./internal/metamorphic -seed 1682318554075133976 -ops "uniform:5000-10000"
```
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMeta.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/internal/manifest/l0_sublevels.go
+++ b/internal/manifest/l0_sublevels.go
@@ -658,14 +658,18 @@ func (s *L0Sublevels) addFileToSublevels(f *FileMetadata, checkInvariant bool) e
 	// Update state in every fileInterval for this file.
 	for i := f.minIntervalIndex; i <= f.maxIntervalIndex; i++ {
 		interval := &s.orderedIntervals[i]
-		if len(interval.files) > 0 &&
-			subLevel <= interval.files[len(interval.files)-1].SubLevel {
+		if len(interval.files) > 0 {
 			if checkInvariant && interval.files[len(interval.files)-1].LargestSeqNum > f.LargestSeqNum {
 				// We are sliding this file "underneath" an existing file. Throw away
 				// and start over in NewL0Sublevels.
 				return errInvalidL0SublevelsOpt
 			}
-			subLevel = interval.files[len(interval.files)-1].SubLevel + 1
+			// interval.files is sorted by sublevels, from lowest to highest.
+			// AddL0Files can only add files at sublevels higher than existing files
+			// in the same key intervals.
+			if maxSublevel := interval.files[len(interval.files)-1].SubLevel; subLevel <= maxSublevel {
+				subLevel = maxSublevel + 1
+			}
 		}
 		interval.estimatedBytes += interpolatedBytes
 		if f.minIntervalIndex < interval.filesMinIntervalIndex {

--- a/internal/manifest/testdata/l0_sublevels
+++ b/internal/manifest/testdata/l0_sublevels
@@ -1683,3 +1683,84 @@ L0.1:     b++++++d evvvvvvvvvvvvvvvj                      rvvvvvvt
 L0.0:  a+++++++++d    fvvvvvvvvvvvvj    l---------o pvvvvvvvvvvvvvvvvvvvvvvvvx
 L6:    a------------------------i          m------------------------------w
        aa bb cc dd ee ff gg hh ii jj kk ll mm nn oo pp qq rr ss tt uu vv ww xx
+
+# Regression test for cockroachdb/cockroach#101896. We must return
+# errInvalidL0SublevelOpt in any case where a new L0 file is being AddL0File'd
+# with a largest sequence number below an existing file in the same interval.
+
+define
+L0
+  000004:a.SET.2-e.SET.3
+  000006:a.SET.7-b.SET.8
+  000007:d.SET.12-f.SET.12
+----
+file count: 3, sublevels: 2, intervals: 5
+flush split keys(2): [b, e]
+0.1: file count: 2, bytes: 512, width (mean, max): 1.5, 2, interval range: [0, 3]
+	000006:[a#7,1-b#8,1]
+	000007:[d#12,1-f#12,1]
+0.0: file count: 1, bytes: 256, width (mean, max): 3.0, 3, interval range: [0, 2]
+	000004:[a#2,1-e#3,1]
+compacting file count: 0, base compacting intervals: none
+L0.1:  a---b    d------f
+L0.0:  a------------e
+       aa bb cc dd ee ff
+
+# Note that 000006 will bump the sublevel for the incoming file to 2. We
+# should still realize that it's slotting below 000007 and return an error.
+
+add-l0-files
+  000015:a.SET.9-g.SET.10
+----
+pebble: L0 sublevel generation optimization cannot be used
+
+# Fully-regenerated L0 sublevels allow us to pick an intra-L0 compaction that
+# does not violate sublevel ordering.
+
+define
+L0
+  000004:a.SET.2-e.SET.3
+  000006:a.SET.7-b.SET.8
+  000007:d.SET.12-f.SET.12
+  000015:a.SET.9-g.SET.10
+----
+file count: 4, sublevels: 4, intervals: 6
+flush split keys(2): [b, e]
+0.3: file count: 1, bytes: 256, width (mean, max): 2.0, 2, interval range: [2, 3]
+	000007:[d#12,1-f#12,1]
+0.2: file count: 1, bytes: 256, width (mean, max): 5.0, 5, interval range: [0, 4]
+	000015:[a#9,1-g#10,1]
+0.1: file count: 1, bytes: 256, width (mean, max): 1.0, 1, interval range: [0, 0]
+	000006:[a#7,1-b#8,1]
+0.0: file count: 1, bytes: 256, width (mean, max): 3.0, 3, interval range: [0, 2]
+	000004:[a#2,1-e#3,1]
+compacting file count: 0, base compacting intervals: none
+L0.3:           d------f
+L0.2:  a------------------g
+L0.1:  a---b
+L0.0:  a------------e
+       aa bb cc dd ee ff gg
+
+# Exclude the d-f file through earliest_unflushed_seqnum.
+
+pick-intra-l0-compaction min_depth=2 earliest_unflushed_seqnum=11
+----
+compaction picked with stack depth reduction 3
+000015,000006,000004
+seed interval: a-b
+L0.3:           d------f
+L0.2:  a++++++++++++++++++g
+L0.1:  a+++b
+L0.0:  a++++++++++++e
+       aa bb cc dd ee ff gg
+
+pick-intra-l0-compaction min_depth=2
+----
+compaction picked with stack depth reduction 3
+000015,000007,000006,000004
+seed interval: a-b
+L0.3:           d++++++f
+L0.2:  a++++++++++++++++++g
+L0.1:  a+++b
+L0.0:  a++++++++++++e
+       aa bb cc dd ee ff gg

